common/src/main/java/io/pravega/common/concurrent/ScheduledQueue.java

```diff
@@ -0,0 +1,350 @@
+            }
+        } else {
+            return ready;
+        }
+    }
+
+    /**
+     * Returns the size of this collection. 
+     * The value is only guaranteed to be accurate if there are not concurrent operations being performed.
+     * If there are, it may reflect the operation or not.
+     * This call is O(1).
+     */
+    @Override
+    public int size() {
+        return (int) Long.min(itemsAdded.get() - itemsRemoved.get(), Integer.MAX_VALUE);
+    }
+
+    /**
+     * Always returns {@code Integer.MAX_VALUE} because
+     * a {@code ScheduledQueue} is not capacity constrained.
+     *
+     * @return {@code Integer.MAX_VALUE}
+     */
+    @Override
+    public int remainingCapacity() {
+        return Integer.MAX_VALUE;
+    }
+
+    /**
+     * Returns an array containing all of the elements in this queue.
+     *
+     * @param a the array into which the elements of the queue are to
+     *          be stored, if it is big enough; otherwise, a new array of the
+     *          same runtime type is allocated for this purpose
+     * @return an array containing all of the elements in this queue
+     */
+    @Override
+    public <T> T[] toArray(@Nonnull T[] a) {
+        ArrayList<E> result = new ArrayList<E>();
+        for (E val : readyTasks) {
+            result.add(val);
+        }
+        for (E val : delayedTasks.values()) {
+            result.add(val);
+        }
+        return result.toArray(a);
+    }
+
+    /**
+     * Removes a single instance of the specified element from this
+     * queue, if it is present, whether or not it has expired.
+     */
+    @Override
+    public boolean remove(Object o) {
+        if (readyTasks.remove(o)) {
+            this.itemsRemoved.incrementAndGet();
+            return true;
+        } else {
+            if (delayedTasks.values().remove(o)) {
+                this.itemsRemoved.incrementAndGet();
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Returns an iterator over all the items in the queue. 
+     *
+     * <p>The returned iterator is
+     * <a href="package-summary.html#Weakly"><i>weakly consistent</i></a>.
+     *
+     * @return an iterator over the elements in this queue
+     */
+    @Override
+    public Iterator<E> iterator() {
+        return Iterators.unmodifiableIterator(Iterators.concat(readyTasks.iterator(), delayedTasks.values().iterator()));
+    }
+
+    @Override
+    public int drainTo(Collection<? super E> c) {
+        return drainTo(c, Integer.MAX_VALUE);
+    }
+
+    @Override
+    public int drainTo(Collection<? super E> c, int maxElements) {
+        int itemCount = 0;
+        blocker.drainPermits();
+        while (itemCount < maxElements) {
+            E item = readyTasks.poll();
+            if (item == null) {
+                break;
+            }
+            c.add(item);
+            itemCount++;
+        }
+        while (itemCount < maxElements) {
```

Comment:
same issue with drainTo here. its behaviour is that if readyTasks are added to after it had drained all elements in it and moved to delayed tasks then it could cause inconsistency.

take following example:
r`i` represents read tasks and d`i` represents delayed tasks.

lets say readyTasks has 1 entry r1.
delayedTasks has 0 entries.


some app could call following:
scheduledQueue.drainTo(2); 
while that is running and has read r1, concurrently following could have been called in another thread:
scheduledQueue.offer(r2);
scheduledQueue.offer(d1);

drainTo method will end up returning r1, d1.
when the order of entry was r1, r2, d1 and d1 was added after r2 so if d1 is drained r2 should have been drained as well.

Reply:
As the inherited javadoc says:
"the behavior of this operation is undefined if the specified collection is modified while the operation is in progress."